File: examples/ios/Examples/OpenCloseRealm.m
// :replace-start: {
//   "terms": {
//     "OpenCloseRealmObjcExamples_": ""
//   }
// }
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

@interface OpenCloseRealmObjcExamples_Task : RLMObject
@end

@implementation OpenCloseRealmObjcExamples_Task
@end

@interface OpenCloseRealmObjc : XCTestCase
@end

@implementation OpenCloseRealmObjc
- (void)testOpenLocalRealm {
    // :code-block-start: open-local-realm
    // Open the default realm
    RLMRealm *defaultRealm = [RLMRealm defaultRealm];

    // Open the realm with a specific file URL, for example a username
    NSString *username = @"GordonCole";
    RLMRealmConfiguration *configuration = [RLMRealmConfiguration defaultConfiguration];
    configuration.fileURL = [[[configuration.fileURL URLByDeletingLastPathComponent]
                             URLByAppendingPathComponent:username]
                             URLByAppendingPathExtension:@"realm"];
    NSError *error = nil;
    RLMRealm *realm = [RLMRealm realmWithConfiguration:configuration
                                                 error:&error];
    // :code-block-end:
    (void)realm;
    (void)defaultRealm;
    
}

- (void)testOpenInMemoryRealm {
    // :code-block-start: open-in-memory-realm
    // Open the realm with a specific in-memory identifier.
    NSString *identifier = @"MyRealm";
    RLMRealmConfiguration *configuration = [[RLMRealmConfiguration alloc] init];
    configuration.inMemoryIdentifier = identifier;
    // Open the realm
    RLMRealm *realm = [RLMRealm realmWithConfiguration:configuration error:nil];
    // :code-block-end:
    (void)realm;
}
    

- (void)testConfigureObjectTypes {
    // :code-block-start: configure-object-types
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    // :remove-start:
    config.inMemoryIdentifier = @"test";
    // :remove-end:
    
    // Given a RLMObject subclass called `OpenCloseRealmObjcExamples_Task`
    // Limit the realm to only the Task object. All other
    // Object- and EmbeddedObject-derived classes are not added.
    config.objectClasses = @[[OpenCloseRealmObjcExamples_Task class]];
    
    NSError *error = nil;
    RLMRealm *realm = [RLMRealm realmWithConfiguration:config error:&error];
    
    if (error != nil) {
        // Something went wrong
    } else {
        // Use realm
    }
    // :code-block-end:
    (void)realm;
}

- (void)testHandleError {
    // :code-block-start: handle-error
    NSError *error = nil;
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    RLMRealm *realm = [RLMRealm realmWithConfiguration:config error:&error];
    if (!realm) {
        // Handle error
        return;
    }
    // Use realm
    // :code-block-end:
}

@end

// :replace-end:

File: examples/ios/Examples/OpenCloseRealm.swift
import XCTest
import RealmSwift

class OpenCloseRealm: XCTestCase {
    func testOpenLocalRealm() {
        // :code-block-start: open-local-realm
        // Open the default realm
        let defaultRealm = try! Realm()

        // Open the realm with a specific file URL, for example a username
        let username = "GordonCole"
        var config = Realm.Configuration.defaultConfiguration
        config.fileURL!.deleteLastPathComponent()
        config.fileURL!.appendPathComponent(username)
        config.fileURL!.appendPathExtension("realm")
        let realm = try! Realm(configuration: config)
        // :code-block-end:
    }

    func testOpenInMemoryRealm() {
        // :code-block-start: open-in-memory-realm
        // Open the realm with a specific in-memory identifier.
        let identifier = "MyRealm"
        let config = Realm.Configuration(
            inMemoryIdentifier: identifier)
        // Open the realm
        let realm = try! Realm(configuration: config)
        // :code-block-end:
    }

    func testConfigureObjectTypes() {
        // :code-block-start: configure-object-types
        var config = Realm.Configuration.defaultConfiguration
        // :remove-start:
        config.inMemoryIdentifier = "test"
        // :remove-end:

        // Given: `class Task: Object`
        // Limit the realm to only the Task object. All other
        // Object- and EmbeddedObject-derived classes are not added.
        config.objectTypes = [Task.self]

        let realm = try! Realm(configuration: config)
        // :code-block-end:
    }

    func testTvOs() {
        // :code-block-start: tvos-share-path
        let fileUrl = FileManager.default
            .containerURL(forSecurityApplicationGroupIdentifier: "group.com.mongodb.realm.examples.extension")!
            .appendingPathComponent("Library/Caches/default.realm")
        // :code-block-end:
        print(fileUrl)
    }

    func testHandleError() {
        // :code-block-start: handle-error
        do {
            let realm = try Realm()
            // Use realm
        } catch let error as NSError {
            // Handle error
        }
        // :code-block-end:
    }
}
